thunder/core/dtypes.py

```diff
@@ -80,23 +81,30 @@ def is_weak(self):
         return self._is_weak
 
     def shortname(self):
-        return f"{self._shortname}{8 * self._bytes}"
+        return f"{self._shortname}{8 * self._bytes}{f'_{self._variant}' if self._variant else ''}"
 
     # TODO Fix name printing
     def __repr__(self):
-        return f"{self._name}{8 * self._bytes}{'_' if self._is_weak else ''}"
+        return (
+            f"{self._name}{8 * self._bytes}{f'_{self._variant}' if self._variant else ''}{'_' if self._is_weak else ''}"
+        )
 
     def __str__(self):
         return self.__repr__()
 
     def __hash__(self) -> int:
-        return hash((self._name, self._bytes, self._is_weak))
+        return hash((self._name, self._bytes, self._is_weak, f"{self._variant if self._variant else ''}"))
```

Comment:
I think `hash(None)` should work so this can probably written as `hash((self._name, self._bytes, self._is_weak, self._variant))`

Reply:
I think we should revert it back to `f"{self._variant if self._variant else ''}"` as hash of empty string is always `0`. 

I have verified that `hash(None)` is not same across different Python process. This is maybe the reason why the pytest sharding is failing. A PR fixed this to be constant across different processes https://github.com/python/cpython/pull/99541 but as per the comment https://github.com/python/cpython/pull/99541#issuecomment-1974098703, it is not guaranteed to be constant by the language spec.

My bad 😓!